Fixed #32789 -- Made feeds emit elements with no content as self-closing tags.

The-Daishogun opened the pull request:
Hello 
this was a problem i encountered at work where the syndication feed doesn't allow self closing tags. for example it produces tags like ``` <enclosure url="http://someurl.com"></enclosure>``` instead of the standard ```<enclosure url="http://someurl.com" />``` i suspect this is caused by the fact that when this code was written self closing tags didn't exist yet. but it has been supported in [XMLGenerator](https://docs.python.org/3/library/xml.sax.utils.html#xml.sax.saxutils.XMLGenerator) since python 3.2 and since it was such a small addition i added them myself.
i appreciate any feedback.
thank you for your time 

ticket-32789

hannseman commented on django/utils/feedgenerator.py:
```diff
@@ -187,7 +188,10 @@ class RssFeed(SyndicationFeed):
 
     def write(self, outfile, encoding):
-        handler = SimplerXMLGenerator(outfile, encoding)
+        if sys.version_info.major == 3 and sys.version_info.minor >= 2:
```
The minimum supported python version for the next Django release is 3.8 so this version check is not needed.

hannseman reviewed (commented):
Thanks! All non-trivial changes (typos etc) require an [accepted ticket](https://code.djangoproject.com/newticket).

The-Daishogun commented:
> Thanks! All non-trivial changes (typos etc) require an [accepted ticket](https://code.djangoproject.com/newticket).

Hi. the [ticket link](https://code.djangoproject.com/ticket/32789)

smithdc1 commented:
Thanks for the PR. Could you add a test for this? 

Also what do you think about backwards compatibility implications of this change?

felixxm commented:
> Also what do you think about backwards compatibility implications of this change?

I would add a short note in the _"Backwards incompatible changes in 4.0 -> Miscellaneous"_, e.g.
```
* :class:`~django.utils.feedgenerator.RssFeed` and its subclasses now emit
  elements with no content as a self-closed tags.
```

The-Daishogun commented:
> Thanks for the PR. Could you add a test for this?
> 
> Also what do you think about backwards compatibility implications of this change?

thank you for your time. 
I added a simple test which checks the if the enclosure tag is self closing or not. since the change happens in SimplerXMLGenerator any tags without text between the opening and closing tags will become self closing. so by checking the enclosures we can get a good idea if the whole thing works or not, plus feed validator websites like [W3](https://validator.w3.org/feed/ )  won't validate the feed if the enclosures are not self closing so i think this is the best place to test it out.

regarding backwards compatibility implications this might have, per @hannseman 's comment about the next major release dropping support for python 3.8 and below i don't think there will be any issues since the short_empty_elements parameter was added in python 3.2 ( [XMLGenerator Documentation](https://docs.python.org/3/library/xml.sax.utils.html#xml.sax.saxutils.XMLGenerator)).

I also added the parameter to Atom1Feed since they also have self closing tags such as ```<link href="#" />``` as welll. [Atom1Feed](https://validator.w3.org/feed/docs/atom.html)

felixxm commented:
> I also added the parameter to Atom1Feed since they also have self closing tags such as `<link href="#" />` as welll. [Atom1Feed](https://validator.w3.org/feed/docs/atom.html)

This should also be tested.

The-Daishogun commented:
> > Also what do you think about backwards compatibility implications of this change?
> 
> I would add a short note in the _"Backwards incompatible changes in 4.0 -> Miscellaneous"_, e.g.
> 
> ```
> * :class:`~django.utils.feedgenerator.RssFeed` and its subclasses now emit
>   elements with no content as a self-closed tags.
> ```

> > I also added the parameter to Atom1Feed since they also have self closing tags such as `<link href="#" />` as welll. [Atom1Feed](https://validator.w3.org/feed/docs/atom.html)
> 
> This should also be tested.



both have been added.

felixxm reviewed (commented):
@The-Daishogun Thanks for updates :+1: Added tests work without this patch because `Feed` calls [.write()](https://docs.python.org/3/library/xml.etree.elementtree.html?highlight=short_empty_elements#xml.etree.ElementTree.ElementTree.write) which emits self-closed tags by default.

I'm not sure if we need this change :thinking: It looks unnecessary.

felixxm commented:
> @The-Daishogun Thanks for updates +1 Added tests work without this patch because `Feed` calls [.write()](https://docs.python.org/3/library/xml.etree.elementtree.html?highlight=short_empty_elements#xml.etree.ElementTree.ElementTree.write) which emits self-closed tags by default.
> 
> I'm not sure if we need this change thinking It looks unnecessary.

OK my mistake :facepalm: , I mixed `Feed.write()` with `XMLGenerator.write()` etc. this patch is desirable. However, tests are incorrect because we don't want to test the `minidom` behavior but `RssFeed` and  `Atom1Feed` behavior. I would add tests similar to the example in the [`django.utils.feedgenerator`](https://docs.djangoproject.com/en/3.2/ref/utils/#module-django.utils.feedgenerator) docs but with a string stream instead of the file stream.

felixxm commented on docs/releases/4.0.txt:
```diff
@@ -432,6 +432,9 @@ Miscellaneous
 * The ``object`` argument of undocumented ``ModelAdmin.log_addition()``,
   ``log_change()``, and ``log_deletion()`` methods is renamed to ``obj``.
 
+* :class:`~django.utils.feedgenerator.RssFeed` and `~django.utils.feedgenerator.Atom1Feed` and their subclasses now emit
+  elements with no content as a self-closed tags.
```
Wrap at 79 chars.

The-Daishogun replied:
thanks for the feedback. totally missed these two.

felixxm reviewed (approved):
@The-Daishogun Thanks :+1: Welcome aboard :boat: 

I squashed commits and simplified tests.

The-Daishogun commented:
> @The-Daishogun Thanks 👍 Welcome aboard ⛵
> 
> I squashed commits and simplified tests.

thank you so much for your feedback. i hope i can continue contributing to this great project which has made my life easier so many times 😄
